Add cw20 usage documentation (#73)
InterWasm/docs: learn/using-contracts/cw20.md

---
title: Sending CW20 Tokens
order: 2
---

Now that you have set up your client, let's get going and try out the
[`cw20-base`](https://github.com/CosmWasm/cosmwasm-plus/tree/master/contracts/cw20-base)
token contract, which implements the
[`cw20` spec](https://github.com/CosmWasm/cosmwasm-plus/blob/master/packages/cw20/README.md).

We will show how to upload the compiled wasm code, instantiate your own contract (possibly
reusing code), and then mint and transfer tokens on that contract.

## Upload and Instantiate a Contract

I will walk you though how to set up an example CW20 contract on CoralNet.

### Example: STAR

The first contract I uploaded was STAR tokens, or "Golden Stars" to be distribute to the
[first 7 validators](https://bigdipper.coralnet.cosmwasm.com/validators) on the network.

Please do not copy this verbatum, but look to see how such a contract is setup and deployed the first time.

```ts
const client = await useOptions(coralnetOptions).setup(YOUR_PASSWORD_HERE);

const cw20 = CW20(client);
const codeId = await cw20.upload();
console.log(`CodeId: ${codeId}`);
// output: 4

> .editor

Reviewer: ```suggestion
// enable REPL editor mode to edit multiline code then execute
.editor
```

const initMsg: InitMsg = {
  name: "Golden Stars",
  symbol: "STAR",
  decimals: 2,
  // list of all validator self-delegate addresses - 100 STARs each!
  initial_balances: [
    { address: "coral1exta8hzrghyt5umd4jh55kfkmp0tv3hyg8krc5", amount: "10000"},
    { address: "coral13mcejut8e5tncs59zcs4yn4envcd98vx682frk", amount: "10000"},
    { address: "coral10zn0d2eeust0495crtr3zqz7t688hg0s53afrh", amount: "10000"},
    { address: "coral1qvrcashqpemlkhrqphzv9n5nutdxpafmdefgcl", amount: "10000"},
    { address: "coral14f8nvyy4c9pyn78dgv0k6syek3jjjrkyz747kj", amount: "10000"},
    { address: "coral1e86v774dch5uwkks0cepw8mdz8a9flhhapvf6w", amount: "10000"},
    { address: "coral1hf50trj7plz2sd8cmcvn7c8ruh3tjhc2nhyl7l", amount: "10000"},
  ],
  mint: {
    minter: client.senderAddress,
  },
};

> ^D

Reviewer: ```suggestion
// exit editor using `^D` and execute entered code 
^D
```

const contract = await cw20.instantiate(codeId, initMsg, "STAR");
console.log(`Contract: ${contract.contractAddress}`);
// Contract: coral16t7y0vrtpqjw2d7jvc2209yan9002339mg4mrv

console.log(await contract.balance("coral13mcejut8e5tncs59zcs4yn4envcd98vx682frk"));
// 10000
console.log(await contract.balance());
// 0
```

If you want to see this is real, just check out
[Code ID 4](https://coralnet.wasm.glass/#/codes/4)
and [contract `coral16t7y0vrtpqjw2d7jvc2209yan9002339mg4mrv`](https://coralnet.wasm.glass/#/contracts/coral16t7y0vrtpqjw2d7jvc2209yan9002339mg4mrv)
on the wasm.glass code explorer.

### Try this at home: MINE

> Now that we have that uploaded, we can easily make a second contract. This one, please

Reviewer: I found great vuepress gems today. https://v1.vuepress.vuejs.org/guide/markdown.html#custom-containers. We can use warning, tip containers to point out warning. I'll go through the documentation after this PR and add containers.

Author: Ah, cool. Happy for any design improvements.

I would merge this PR once I address the comments, then let you add another PR to make it cleaner.

const mine = await cw20.instantiate(codeId, initMsg, "MINE");
console.log(`Contract: ${mine.contractAddress}`);
// Contract: coral13nkgqrfymug724h8pprpexqj9h629sa3yzcqfk

// now, check the configuration
mine.balance();
mine.token_info()
mine.minter()
```

Look, you're rich now! Time to share the wealth.

## Using a contract

In this section, we will show you how to use your newly constructed token.
You can keep typing along in the same REPL that you used to create the `MINE`
tokens (or whatever better name you invented), but if you closed it down and
come back, here's how to re-connect:

```ts
const client = await useOptions(coralnetOptions).setup(YOUR_PASSWORD_HERE);
const cw20 = CW20(client);

// if you forgot your address, but remember your label, you can find it again
const contracts = await client.getContracts(4)
contracts
const contractAddress = contracts.filter(x => x.label === 'MINE')[0].address;

// otherwise, you can just cut and paste from before
const contractAddress = "coral13nkgqrfymug724h8pprpexqj9h629sa3yzcqfk"

// now, connect to that contract and make sure it is yours
const mine = cw20.use(contractAddress);
mine.token_info()
mine.minter()
mine.balance()
```

> Okay, you are connected to your contract. Let's DO SOMETHING.

Reviewer: ```suggestion
Okay, you are connected to your contract. Let's see what cw20 is capable of.
```

mine.balance(other)
mine.balance()
```

Great, you are moving stuff around and see it in your queries and in the block explorer.
Time to act like a pro.

## Interactive Discovery

So far you have been cut-and-pasting commands I have been telling you. But what else
can you do with this contract? Luckily Javascript has some nice introspection. And we
added some extra bonuses there. Just type `mine` in the REPL and see the list of methods:

```
>> mine
{ contractAddress: 'coral13nkgqrfymug724h8pprpexqj9h629sa3yzcqfk',
  balance: [AsyncFunction: balance],
  allowance: [AsyncFunction: allowance],
  token_info: [AsyncFunction: token_info],
  minter: [AsyncFunction: minter],
  mint: [AsyncFunction: mint],
  transfer: [AsyncFunction: transfer],
  burn: [AsyncFunction: burn],
  increaseAllowance: [AsyncFunction: increaseAllowance],
  decreaseAllowance: [AsyncFunction: decreaseAllowance],
  transferFrom: [AsyncFunction: transferFrom] }
```

But how do I call them? What arguments do they take?
There is a great `.type` operator. Unfortunately it doesn't work
with members directly, but with a local variable, we can get the info.

> This is not what we want:

Reviewer: Telling negative results before showing code is a bit confusing. The reader will be like: **then why do you show this to me?**. Swap the lines of the code example below and this line.

Author: Good idea. After doing the proper way, you will think "why so long, this is a shortcut".
I tried the first one first, but yeah, I guess you want the answer first
